### PR TITLE
Subworkflow for organelles

### PR DESCRIPTION
**Pull request by ksenia-krasheninnikova**
Implements assembling mito from raw reads and from assembly

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/genomeassembly/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the sanger-tol/genomeassembly _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Review by priyanka-surana — COMMENTED** [review]
Let's test this after merging the other two PRs and then good to move forward. You might want to add associated documentation to this PR.

File: modules/local/mitohifi.nf
[the remark this answers is not in the document]
Comment (ksenia-krasheninnikova) [reply modules/local/mitohifi.nf:1]: Yes, this is the one

File: subworkflows/local/organelles.nf [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                        code)    
+    }
+    if ( contigs ) {
+        MITOHIFI_MITOHIFI_ASM( contigs.map{meta, contigs -> [meta, [], contigs]}, 
+                        MITOHIFI_FINDMITOREFERENCE.out.fasta,
+                        MITOHIFI_FINDMITOREFERENCE.out.gb,
+                        code)    
+    }
+    
+    emit:
```
[the remark this answers is not in the document]
Comment (ksenia-krasheninnikova) [reply subworkflows/local/organelles.nf:37]: Seems it looks better now.

**Review by priyanka-surana — COMMENTED** [review]
Do you want to make the changes and merge this PR? And do the changes for CI testing in a separate PR?

**Comment by ksenia-krasheninnikova** [comment]
Removed all redundant files from the emit section. I think it can be merged in now. Thanks @priyanka-surana

File: nextflow.config [hunk 2]
```diff
@@ -19,6 +19,7 @@ params {
 
     // Assembly options
     hifiasm_hic_on             = false
+    organelles_on              = true
```
Comment (priyanka-surana) [inline nextflow.config:22]: We are recommended to keep the default as `false` and then turn on via command line flags or the config.

File: subworkflows/local/organelles.nf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                        MITOHIFI_FINDMITOREFERENCE.out.fasta,
+                        MITOHIFI_FINDMITOREFERENCE.out.gb,
+                        code)    
+        mitohifi_asm = MITOHIFI_MITOHIFI_ASM.out
+
+    }
+    
+    emit:
+
+    versions = ch_versions
```
Comment (priyanka-surana) [inline subworkflows/local/organelles.nf:40]: Now there are no output files. Is this intentional?

**Review by priyanka-surana — COMMENTED** [review]
A couple minor changes requested.

File: nextflow.config (hunk 2 above)
Comment (ksenia-krasheninnikova) [reply nextflow.config]: ok

File: subworkflows/local/organelles.nf (hunk 3 above)
Comment (ksenia-krasheninnikova) [reply subworkflows/local/organelles.nf:40]: While we need this files in the final output, we don't need them down the workflow. Initially I thought the emit section has to do something with the files being published, but actually it works fine the way it is now.

**Comment by muffato** [comment]
> I find this `emit` section confusing and difficult to navigate. Could be just me since I am not familiar with MitoHifi. Looking for a more informed view @muffato since you are familiar with the tool.

I'm not that familiar with MitoHifi :) !

I can see you cleaned up the entire `emit` section. It simplifies things a lot ! For the record, I had some questions about that structure:

1. Could you make the MitoHifi _module_ take a single input like this:
    1. Define it as `tuple val(meta), path(input)` rather than the mutually-exclusive `reads` and `contigs`
    2. Use `$args $input` in the MitoHifi command-line rather than `${run_type}`
    3. First check that `$args` does contain either `-c` or `-r` (one of them is mandatory)
2. Then, do you need this one subworkflow to choose between the assembly and read mode and duplicate the input and output channels ? Could you imagine having a single subworkflow that is agnostic of the mode (just one input channel and one et of output channels), and in the parent workflow you would import it twice: once as `ORGANELLES_READS` and once as `ORGANELLES_ASM`

**Comment by ksenia-krasheninnikova** [comment]
> 
> 1. Could you make the MitoHifi _module_ take a single input like this:
>    
>    1. Define it as `tuple val(meta), path(input)` rather than the mutually-exclusive `reads` and `contigs`
>    2. Use `$args $input` in the MitoHifi command-line rather than `${run_type}`
>    3. First check that `$args` does contain either `-c` or `-r` (one of them is mandatory)
> 2. Then, do you need this one subworkflow to choose between the assembly and read mode and duplicate the input and output channels ? Could you imagine having a single subworkflow that is agnostic of the mode (just one input channel and one et of output channels), and in the parent workflow you would import it twice: once as `ORGANELLES_READS` and once as `ORGANELLES_ASM`

Thank you @muffato for suggestions. I've updated the code, could you have a look if it's what you mean please? I had to put -r/-c into ext.args2 because the input file should follow the corresponding option immediately:

```
usage: MitoHiFi [-h] (-r <reads>.fasta | -c <contigs>.fasta) -f
                <relatedMito>.fasta -g <relatedMito>.gbk -t <THREADS> [-d]
                [-a {animal,plant,fungi}] [-p <PERC>] [-m <BLOOM FILTER>]
                [--max-read-len MAX_READ_LEN] [--mitos]
                [--circular-size CIRCULAR_SIZE]
                [--circular-offset CIRCULAR_OFFSET] [-winSize WINSIZE]
                [-covMap COVMAP] [-v] [-o <GENETIC CODE>]
```

The purpose of the original implementation of the organelles workflow was to separate the whole organelles logic into the separate workflow to lighten up the main workflow. My concern though is that the main workflow getting rather long. Is it a standard practice to keep it that way? It will be come more readable when I add documentation.

**Review by muffato — COMMENTED** [review]
I like this. I find it much more straightforward !
Can you propose your updated mitohifi/mitohifi module to nf-core (can do that after this PR) and suggest _mahesh-panchal_ and _scorreard_ for review ? They both work on genome-assembly pipelines.

File: modules/local/get_calcuts_params.nf [hunk 4]
```diff
@@ -4,7 +4,7 @@ process GET_CALCUTS_PARAMS {
     conda "conda-forge::python"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/python:3' :
-        'python:3' }"
+        'biocontainers/python:3.9' }"
```
[the remark this answers is not in the document]
Comment (ksenia-krasheninnikova) [reply modules/local/get_calcuts_params.nf]: Both options work. Changed to docker.io/python:3.9.

**Comment by ksenia-krasheninnikova** [comment]
@muffato 
Merged it updated mitohifi module from nf-core. All the review items must have been addressed for now!
PR ready for the final review. 
Thanks!

**Review by muffato — APPROVED** [review]
As you said, the module is now in sync with nf-core, meaning no local `.diff` file.

All good !

**Pull request merged**